mapdataprefetcher/app/src/main/java/com/here/android/example/map/data/prefetcher/MainActivity.java

```diff
@@ -0,0 +1,150 @@
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        switch (requestCode) {
+            case REQUEST_CODE_ASK_PERMISSIONS: {
+                for (int index = 0; index < permissions.length; index++) {
+                    if (grantResults[index] != PackageManager.PERMISSION_GRANTED) {
+
+                        /*
+                         * If the user turned down the permission request in the past and chose the
+                         * Don't ask again option in the permission request system dialog.
+                         */
+                        if (!ActivityCompat
+                                .shouldShowRequestPermissionRationale(this, permissions[index])) {
+                            Toast.makeText(this, "Required permission " + permissions[index]
+                                            + " not granted. "
+                                            + "Please go to settings and turn on for sample app",
+                                    Toast.LENGTH_LONG).show();
+                        } else {
+                            Toast.makeText(this, "Required permission " + permissions[index]
+                                    + " not granted", Toast.LENGTH_LONG).show();
+                        }
+                    }
+                }
+
+                setupMapFragmentView();
+                break;
+            }
+            default:
+                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        }
+    }
+
+    private void setupMapFragmentView() {
+        // All permission requests are being handled. Create map fragment view. Please note
+        // the HERE Mobile SDK requires all permissions defined above to operate properly.
+        m_mapFragmentView = new MapFragmentView(this);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.map_menu, menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
+        if (m_mapFragmentView.isRouteSimulationFinished) {
+            if (!item.isChecked()) {
+                item.setChecked(true);
+
+                switch (item.getItemId()) {
+                    case BOUNDING_BOX:
+                        m_mapFragmentView.setMapType(MapType.MAP_BOUNDING_BOX);
+                        return true;
+                    case ROUTE:
+                        m_mapFragmentView.setMapType(MapType.ROAD_RADIUS);
+                        return true;
+                }
+            }
+        } else {
+            Toast.makeText(this, "You can't change map when simulation is running.",
+                    Toast.LENGTH_LONG).show();
+        }
+        return true;
```

Comment:
Move this logic into `MapFragmentView` class, so it will look like `m_mapFragmentView.onOptionsItemSelected(...)`